services/oauth/v2/controlplane/cp_connector.go

```diff
@@ -0,0 +1,186 @@
+package controlplane
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/services/oauth/v2/common"
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+)
+
+var ErrTypMap = map[syscall.Errno]string{
+	syscall.ECONNRESET:   "econnreset",
+	syscall.ECONNREFUSED: "econnrefused",
+	syscall.ECONNABORTED: "econnaborted",
+	syscall.ECANCELED:    "ecanceled",
+}
+
+type ControlPlaneConnector interface {
+	CpApiCall(cpReq *ControlPlaneRequest) (int, string)
+}
+
+type controlPlaneConnector struct {
+	client     HttpClient
+	logger     logger.Logger
+	timeOut    time.Duration
+	loggerName string
+}
+
+func NewControlPlaneConnector(conf *config.Config, options ...func(*controlPlaneConnector)) ControlPlaneConnector {
+	cpConnector := &controlPlaneConnector{
+		client: &http.Client{
+			Transport: http.DefaultTransport,
+			Timeout:   conf.GetDuration("HttpClient.oauth.timeout", 30, time.Second),
+		},
+		logger: logger.NewLogger().Child("ControlPlaneConnector"),
+	}
+
+	for _, opt := range options {
+		opt(cpConnector)
+	}
+
+	return cpConnector
+}
+
+/*
+WithClient is a functional option to set the client for the ControlPlaneConnector
+*/
+func WithClient(client HttpClient) func(*controlPlaneConnector) {
+	return func(cpConn *controlPlaneConnector) {
+		cpConn.client = client
+	}
+}
+
+/*
+WithParentLogger is a functional option to set the parent logger for the ControlPlaneConnector
+*/
+func WithLogger(parentLogger logger.Logger) func(*controlPlaneConnector) {
+	return func(cpConn *controlPlaneConnector) {
+		cpConn.logger = parentLogger
+	}
+}
+
+/*
+WithCpClientTimeout is a functional option to set the timeout for the ControlPlaneConnector
+*/
+func WithCpClientTimeout(timeout time.Duration) func(*controlPlaneConnector) {
+	return func(h *controlPlaneConnector) {
+		h.timeOut = timeout
+	}
+}
+
+/*
+processResponse is a helper function to process the response from the control plane
+*/
+func processResponse(resp *http.Response) (statusCode int, respBody string) {
+	var respData []byte
+	var ioUtilReadErr error
+	if resp != nil && resp.Body != nil {
+		respData, ioUtilReadErr = io.ReadAll(resp.Body)
+		if ioUtilReadErr != nil {
+			return http.StatusInternalServerError, ioUtilReadErr.Error()
+		}
+	}
+	// Detecting content type of the respData
+	contentTypeHeader := strings.ToLower(http.DetectContentType(respData))
+	// If content type is not of type "*text*", overriding it with empty string
+	if !(strings.Contains(contentTypeHeader, "text") ||
+		strings.Contains(contentTypeHeader, "application/json") ||
+		strings.Contains(contentTypeHeader, "application/xml")) {
```

Comment:
[optional]: Using regex will probably be a more cleaner approach here:

```suggestion
	if !contentTypePattern.MatchString(contentTypeHeader) {
```

regex can be compiled once and set at the package level.
```GO
var (
	contentTypePattern = regexp.MustCompile(`text|application/json|application/xml`)
)
```